yt/frontends/halo_catalog/data_structures.py

```diff
@@ -146,3 +177,363 @@ def _is_valid(self, *args, **kwargs):
+
+    def _get_halo_scalar_index(self, ptype, identifier):
+        i_scalar = self._get_halo_file_indices(ptype, [identifier])[0]
+        scalar_index = identifier - self._halo_index_start[ptype][i_scalar]
+        return scalar_index
+
+    def _get_halo_values(self, ptype, identifiers, fields, f=None):
+        """
+        Get field values for halo data containers.
+        """
+
+        # if a file is already open, don't open it again
+        filename = None if f is None else f.filename
+
+        data = defaultdict(lambda: np.empty(identifiers.size))
+        i_scalars = self._get_halo_file_indices(ptype, identifiers)
+        for i_scalar in np.unique(i_scalars):
+            target = i_scalars == i_scalar
+            scalar_indices = identifiers - self._halo_index_start[ptype][i_scalar]
+
+            # only open file if it's not already open
+            my_f = (
+                f
+                if self.data_files[i_scalar].filename == filename
+                else h5py.File(self.data_files[i_scalar].filename, "r")
+            )
+
+            for field in fields:
+                data[field][target] = self._read_halo_particle_field(
+                    my_f, ptype, field, scalar_indices[target]
+                )
+
+            if self.data_files[i_scalar].filename != filename:
+                my_f.close()
+
+        return data
+
+    def _identify_base_chunk(self, dobj):
+        pass
+
+    def _read_halo_particle_field(self, fh, ptype, field, indices):
+        return fh[field][indices]
+
+    def _read_particle_fields(self, fields, dobj, chunk=None):
+        if len(fields) == 0:
+            return {}, []
+        fields_to_read, fields_to_generate = self._split_fields(fields)
+        if len(fields_to_read) == 0:
+            return {}, fields_to_generate
+        fields_to_return = self.io._read_particle_selection(dobj, fields_to_read)
+        return fields_to_return, fields_to_generate
+
+    def _setup_data_io(self):
+        super(YTHaloParticleIndex, self)._setup_data_io()
+        if self.real_ds._instantiated_index is None:
+            self.real_ds.index
+
+        # inherit some things from parent index
+        for attr in ["data_files", "total_particles"]:
+            setattr(self, attr, getattr(self.real_ds.index, attr))
+
+        self._calculate_particle_index_starts()
+        self._create_halo_id_table()
+
+
+class HaloDataset(ParticleDataset):
+    """
+    Base class for dataset accessing particles from halo catalogs.
+    """
+
+    def __init__(self, ds, dataset_type):
+        self.real_ds = ds
+        for attr in [
+            "filename_template",
+            "file_count",
+            "particle_types_raw",
+            "particle_types",
+            "periodicity",
+        ]:
+            setattr(self, attr, getattr(self.real_ds, attr))
+
+        super(HaloDataset, self).__init__(self.real_ds.parameter_filename, dataset_type)
+
+    def print_key_parameters(self):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError
```

Reply:
This is called by the base class and I want to do nothing since this is basically a hidden dataset.